src/librustc_unicode/char.rs

```diff
@@ -470,33 +740,63 @@ impl char {
         ToLowercase(CaseMappingIter::new(conversions::to_lower(self)))
     }
 
-    /// Converts a character to its uppercase equivalent.
+    /// Returns an iterator that yields the uppercase equivalent of a `char`.
     ///
-    /// This performs complex unconditional mappings with no tailoring:
-    /// it maps one Unicode character to its uppercase equivalent
-    /// according to the Unicode database [1]
-    /// and the additional complex mappings [`SpecialCasing.txt`].
-    /// Conditional mappings (based on context or language) are not considered here.
+    /// If no conversion is possible then an iterator with just the input character is returned.
     ///
-    /// A full reference can be found here [2].
+    /// This performs complex unconditional mappings with no tailoring: it maps
+    /// one Unicode character to its uppercase equivalent according to the
+    /// [Unicode database] and the additional complex mappings
+    /// [`SpecialCasing.txt`]. Conditional mappings (based on context or
+    /// language) are not considered here.
     ///
-    /// # Return value
+    /// For a full reference, see [here][reference].
     ///
-    /// Returns an iterator which yields the characters corresponding to the
-    /// uppercase equivalent of the character. If no conversion is possible then
-    /// an iterator with just the input character is returned.
-    ///
-    /// [1]: ftp://ftp.unicode.org/Public/UNIDATA/UnicodeData.txt
+    /// [Unicode database]: ftp://ftp.unicode.org/Public/UNIDATA/UnicodeData.txt
     ///
     /// [`SpecialCasing.txt`]: ftp://ftp.unicode.org/Public/UNIDATA/SpecialCasing.txt
     ///
-    /// [2]: http://www.unicode.org/versions/Unicode7.0.0/ch03.pdf#G33992
+    /// [reference]: http://www.unicode.org/versions/Unicode7.0.0/ch03.pdf#G33992
     ///
     /// # Examples
     ///
+    /// Basic usage:
+    ///
     /// ```
-    /// assert_eq!(Some('C'), 'c'.to_uppercase().next());
+    /// let c = 'c';
+    /// assert_eq!(c.to_uppercase().next(), Some('C'));
+    ///
+    /// // 日本語 does not have case
+    /// let c = '山';
+    /// assert_eq!(c.to_uppercase().next(), Some('山'));
     /// ```
+    ///
+    /// In Turkish, the equivalent of 'i' in Latin has five forms instead of two:
+    ///
+    /// * 'Dotless': I / ı, sometimes written ï
+    /// * 'Dotted': İ / i
+    ///
+    /// Note that the lower case dotted 'i' is the same as the Latin. Therefore:
+    ///
+    /// ```
+    /// let i = 'i';
+    ///
+    /// let upper_i = i.to_uppercase().next();
+    /// ```
+    ///
+    /// The value of `upper_i` here relies on the locale: if we're in `en-US`, it
+    /// should be `Some('I')`, but if we're in `tr_TR`, it should be `Some('İ')`.
```

Comment:
I’d prefer avoiding the word "locale". At least in my mind, it refers to some global state of the environment where the program is running. (On Unix, its `std::env::var("LANG")`.) This is not necessarily the language of a given piece of text that the program is manipulating, particularly in a client/server app where one server-side worker might be serving many users who might speak different languages. (I think it would be a mistake to design APIs that implicitly rely on the current locale.)

Maybe use "the language of the text" or "the text language" instead.